### PR TITLE
Annotate System.Resources.ResourceWriter for nullable ref types

### PR DESCRIPTION
**Pull request by danmoseley**
Contributes to https://github.com/dotnet/corefx/issues/40623

### CODE REVIEW

File: src/Common/src/System/Resources/ResourceWriter.cs [hunk 1]
```diff
@@ -236,8 +224,8 @@ private void Dispose(bool disposing)
                 }
             }
 
-            _output = null;
-            _caseInsensitiveDups = null;
+            _output = null!;
```
Comment (danmosemsft) [inline src/Common/src/System/Resources/ResourceWriter.cs]: In the common case like here where the pattern is null iff disposed, it makes sense to me to bang the assignments in dispose. That way all the rest of the code can get the benefits of the null checking, and no code should run after dispose -- indeed checks are made against null to enforce that.

File: src/Common/src/CoreLib/System/Resources/ResourceTypeCode.cs [hunk 2]
```diff
@@ -2,17 +2,6 @@
...
-**
-**
-**
-**
-**
-** Purpose: Marker for types in .resources files
-**
-**
-===========================================================*/
-
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/Resources/ResourceTypeCode.cs]: Why edit this file?

File: src/Common/src/System/Resources/ResourceWriter.cs [hunk 3]
```diff
@@ -2,17 +2,7 @@
...
-** Class:  ResourceWriter
-**
-**
-**
-** Purpose: Default way to write strings to a CLR resource
-** file.
-**
-**
-===========================================================*/
+#nullable enable
```
Comment (stephentoub) [inline src/Common/src/System/Resources/ResourceWriter.cs]: Why is this needed?  Is this file included in another project that doesn't set `<Nullable>enable</Nullable>`?

File: src/Common/src/System/Resources/ResourceWriter.cs [hunk 4]
```diff
@@ -422,7 +412,7 @@ public void Generate()
             bw.Flush();
 
             // Indicate we've called Generate
-            _resourceList = null;
+            _resourceList = null!;
```
Comment (stephentoub) [inline src/Common/src/System/Resources/ResourceWriter.cs]: I agree with the argument you made above for Dispose (we do that in other disposal cases as well), but why is it ok here?

File: src/Common/src/CoreLib/System/Resources/ResourceTypeCode.cs (hunk 2 above)
Comment (danmosemsft) [reply src/Common/src/CoreLib/System/Resources/ResourceTypeCode.cs]: I have been removing such low value comments when I come across them

File: src/Common/src/System/Resources/ResourceWriter.cs (hunk 3 above)
Comment (danmosemsft) [reply src/Common/src/System/Resources/ResourceWriter.cs]: Yes, I think it was S.Resources.Extensions.

File: src/Common/src/System/Resources/ResourceWriter.cs (hunk 4 above)
Comment (buyaa-n) [reply src/Common/src/System/Resources/ResourceWriter.cs]: Seems should be nullable

**Review by buyaa-n — APPROVED** [review]
Other than @stephentoub's comment LGTM

File: src/Common/src/System/Resources/ResourceWriter.cs (hunk 4 above)
Comment (danmosemsft) [reply src/Common/src/System/Resources/ResourceWriter.cs]: Same reasoning -- 
1. `Generate()` is analogous to dispose. After it's been called, the type can no longer be used. 
1. This is the only place it is set to null. It is initialized in all constructors. It is not accessible to derived classes. There are no virtual members.
1. For this reason, all entry points already check for null.

Comment (stephentoub) [reply src/Common/src/System/Resources/ResourceWriter.cs]: > For this reason, all entry points already check for null.

If all entry points check for null, then what's the downside to making the field nullable and reflecting reality?  It looks like you'd need to add a Debug.Assert to one internal function, and everything else would just work. No?

Comment (danmosemsft) [reply src/Common/src/System/Resources/ResourceWriter.cs]: Sure, I can do this. I need to get myself into the nullable annotation mindset I think 😸

**Comment by danmosemsft** [comment]
@stephentoub any more feedabck?

failed test is https://github.com/dotnet/corefx/issues/41911

**Review by stephentoub — APPROVED** [review]
